File: app/src/main/java/me/tagavari/airmessage/util/AnimatingInsetsCallback.java
package me.tagavari.airmessage.util;

import android.graphics.Insets;
import android.os.Build;
import android.view.View;
import android.view.WindowInsets;
import android.view.WindowInsetsAnimation;

import androidx.annotation.NonNull;
import androidx.annotation.RequiresApi;

import java.util.List;

@RequiresApi(api = Build.VERSION_CODES.R)
public class AnimatingInsetsCallback extends WindowInsetsAnimation.Callback implements View.OnApplyWindowInsetsListener {
	private final View view;
	
	private boolean isAnimating = false;
	
	public AnimatingInsetsCallback(View view) {
		super(DISPATCH_MODE_CONTINUE_ON_SUBTREE);
		
		this.view = view;
	}
	
	@Override
	public WindowInsets onApplyWindowInsets(View view, WindowInsets windowInsets) {
		if(isAnimating) return windowInsets;
		Insets insets = windowInsets.getInsets(WindowInsets.Type.systemBars() | WindowInsets.Type.ime());
		view.setPadding(insets.left, view.getPaddingTop(), insets.right, insets.bottom);
		
		return WindowInsets.CONSUMED;
	}
	
	@NonNull
	@Override
	public WindowInsets onProgress(@NonNull WindowInsets windowInsets, @NonNull List<WindowInsetsAnimation> list) {
		view.setPadding(view.getPaddingLeft(), view.getPaddingTop(), view.getPaddingRight(), windowInsets.getInsets(WindowInsets.Type.ime() | WindowInsets.Type.systemBars()).bottom);
		return WindowInsets.CONSUMED;
	}
	
	@Override
	public void onPrepare(@NonNull WindowInsetsAnimation animation) {
		isAnimating = true;
	}
	
	@Override
	public void onEnd(@NonNull WindowInsetsAnimation animation) {
		isAnimating = false;
	}
}

File: app/src/main/java/me/tagavari/airmessage/util/AnimatingInsetsCallback.kt
package me.tagavari.airmessage.util

import android.os.Build
import android.view.View
import androidx.annotation.RequiresApi
import android.view.WindowInsetsAnimation
import android.view.WindowInsets

@RequiresApi(api = Build.VERSION_CODES.R)
class AnimatingInsetsCallback(private val view: View) : WindowInsetsAnimation.Callback(DISPATCH_MODE_CONTINUE_ON_SUBTREE), View.OnApplyWindowInsetsListener {
	private var isAnimating = false
	
	override fun onApplyWindowInsets(view: View, windowInsets: WindowInsets): WindowInsets {
		if(isAnimating) return windowInsets
		val insets = windowInsets.getInsets(WindowInsets.Type.systemBars() or WindowInsets.Type.ime())
		view.setPadding(insets.left, view.paddingTop, insets.right, insets.bottom)
		
		return WindowInsets.CONSUMED
	}
	
	override fun onProgress(windowInsets: WindowInsets, list: List<WindowInsetsAnimation>): WindowInsets {
		view.setPadding(
			view.paddingLeft,
			view.paddingTop,
			view.paddingRight,
			windowInsets.getInsets(WindowInsets.Type.ime() or WindowInsets.Type.systemBars()).bottom
		)
		
		return WindowInsets.CONSUMED
	}
	
	override fun onPrepare(animation: WindowInsetsAnimation) {
		isAnimating = true
	}
	
	override fun onEnd(animation: WindowInsetsAnimation) {
		isAnimating = false
	}
}